sync.md

```diff
@@ -1,131 +1,153 @@
 // when receiving new blocks from the network
 func (syncer *Syncer) InformNewHead(from PeerID, head TipSet) {
 	switch syncer.syncMode {
 	case Bootstrap:
-		go SyncBootstrap(from, head)
+    // InformNewHead will only get called during PeerSetExpansion during bootstrapping
+		go SyncBootstrap(true)
 	case CaughtUp:
 		go syncer.SyncCaughtUp(blk)
 	}
 }
 
 // SyncBootstrap is used to synchronise your chain when first joining
 // the network, or when rejoining after significant downtime.
-func (syncer *Syncer) SyncBootstrap() {
-	syncer.syncLock.Lock()
-	defer syncer.syncLock.Unlock()
-	syncer.peerHeads[from] = head
-	if len(syncer.peerHeads) < BootstrapPeerThreshold {
-		// not enough peers to sync yet...
-		return
-	}
-
-	selectedHead := selectHead(syncer.peerHeads)
-
-	cur := selectedHead
-	var blockSet BlockSet
-	for head.Height() > 0 {
-		// NB: GetBlocks validates that the blocks are in-fact the ones we
-		// requested, and that they are correctly linked to eachother. It does
-		// not validate any state transitions
-		blks := syncer.bsync.GetBlocks(head, RequestWidth)
-		blockSet.Insert(blks)
-
-		head = blks.Last().Parents()
-	}
-
-	genesis := blockSet.GetByHeight(0)
-	if genesis != syncer.genesis {
-		// TODO: handle this...
-		Error("We synced to the wrong chain!")
-		return
-	}
-
-	// Fetch all the messages for all the blocks in this chain
-	// There are many ways to make this more efficient. For now, do the dumb thing
-	blockSet.ForEach(func(b Block) {
-		// FetchMessages should use bitswap to fetch any messages we don't have locally
-		FetchMessages(b)
-	})
-
-	// Now, to validate some state transitions
-	base := genesis
-	for i := 1; i < selectedHead.Height(); i++ {
-		next := blockSet.GetByHeight(i)
-		if !ValidateTransition(base, next) {
-			// TODO: do something productive here...
-			Error("invalid state transition")
-			return
-		}
-	}
-
-	blockSet.PersistTo(syncer.store)
-	syncer.head = bset.Head()
-	syncer.syncMode = CaughtUp
+func (syncer *Syncer) SyncBootstrap(bool expanding) {
+    syncer.syncLock.Lock()
+    defer syncer.syncLock.Unlock()
+    if expanding && len(syncer.peerHeads) < BootstrapPeerThreshold {
+        // not enough peers to sync yet...
+        return
+    }
+  
+  	syncPeers := syncer.peerSet
+  	syncHeads := syncer.peerHeads
+  	if (!expanding) {
+    		syncPeers = syncer.trustedPeers
+      	syncHeads = syncer.trustedHeads
+  	}
+  	// Will now get heaviest head from all the heads from our Peerset
+    selectedHead := selectHead(syncer.peerHeads)
+    
+    cur := selectedHead
+    var blockSet BlockSet
+    for head.Height() > 0 {
+        // NB: GetBlocks validates that the blocks are in-fact the ones we
+        // requested, and that they are correctly linked to each other. It does
+        // not validate any state transitions
+        blks := syncer.bsync.GetBlocks(head, RequestWidth)
+        blockSet.Insert(blks)
+            
+        head = blks.Last().Parents()
+    }
+        
```

Comment:
you dropped checking if the synced chain links back to the expected genesis block. Thats an important part

Reply:
`SyncBootstrap` only gets called by 
- `InformNewHead`, itself called by `addPeerToSet` --> which checks that the new peer's genesis is GENESIS
- `ExpandPeerSet` upon peer set expansion failure --> in which case we sync up to the `TrustedNodes` all of whose genesis have been checked in `InitialConnect`

So it actually gets checked elsewhere which is why I removed it.

With that said, that is so non-obvious and likely to lead to a regression that I will re-add.

I'd rather not remove the checks elsewhere though as they logically make sense there. The duplication is NBD, comparatively, imo. LMK if you disagree!

Reply:
This is getting a bit too in the weeds for a spec imo: I have fallen into the trap of writing code instead of saying "how" things should work.

Going to use an assertion here instead just to make clear that we didn't forget.

Reply:
@sternhenri It doesnt matter if the peer claims that X is their genesis block. They can still claim that, and then give us a completely different chain.

Reply:
That's a v good point. Well taken!

With that said, does [l114 in syncer](https://github.com/filecoin-project/specs/pull/196/files#diff-3d25bc8d97eb696d79b9ab02ade7a634R114) not cover it?